src/futures/spi.rs

```diff
@@ -0,0 +1,67 @@
+//! Serial Peripheral Interface
+
+use core::future::Future;
+
+/// Async transfer
+pub trait Transfer<W: 'static> {
+    /// Error type
+    type Error;
+
+    /// Associated future for the `transfer` method.
+    type TransferFuture<'a>: Future<Output = Result<(), Self::Error>> + 'a
+    where
+        Self: 'a;
+
+    /// Writes `words` to the slave from the `write` buffer. Puts the words returned in the `read` buffer.
+    /// This method uses separate `write` and `read` buffers.
+    fn transfer<'a>(&'a mut self, write: &'a [W], read: &'a mut [W]) -> Self::TransferFuture<'a>;
+}
+
+/// Async transfer in place.
+pub trait TransferInPlace<W: 'static> {
+    /// Error type
+    type Error;
+
+    /// Associated future for the `transfer_inplace` method.
+    type TransferInPlaceFuture<'a>: Future<Output = Result<(), Self::Error>> + 'a
+    where
+        Self: 'a;
+
+    /// Writes `words` to the slave from the `readwrite` buffer and reads words into the same buffer.
+    /// This method uses a single `readwrite` buffer.
+    ///
+    /// The returned buffer is the initialized `readwrite` buffer.
+    fn transfer_inplace<'a>(&'a mut self, readwrite: &'a mut [W]) -> Self::TransferInPlaceFuture<'a>;
```

Comment:
The `readwrite` buffer should probably be named `words` or similar, so that no operation order is implied.